src/wallet/rpcdump.cpp

```diff
@@ -324,6 +324,98 @@ RPCHelpMan importaddress()
     };
 }
 
+static void RemoveAddress(CWallet*, const CTxDestination&, bool);
+static void RemoveScript(CWallet* const pwallet, const CScript& script, bool is_redeem_script, bool purge_txns) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    isminetype ismine = pwallet->IsMine(script);
+    if (!is_redeem_script && ismine == ISMINE_SPENDABLE) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "The wallet contains the private key for this address");
+    } else if (!is_redeem_script && ismine == ISMINE_NO) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "The wallet does not contain this address or script");
+    }
+    CHECK_NONFATAL(ismine == ISMINE_WATCH_ONLY || is_redeem_script);
+
+    LegacyScriptPubKeyMan& spk_man = EnsureLegacyScriptPubKeyMan(*pwallet);
+
+    if (spk_man.HaveWatchOnly(script) && !spk_man.PurgeWatchOnly(script)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error removing address/script from wallet");
+    }
+    if (is_redeem_script) {
+        CScriptID scriptID = CScriptID(script);
+        if (!spk_man.HaveCScript(scriptID)) {
```

Comment:
In commit "rpc: Added ability to remove watch only addresses" (b8eb5880693358951229512987beffcff51e0ecf)

Important: Because there is no RemoveCScript function being called here to undo the effect of AddCScript/AddCScriptWithDB, the `removeaddress` call does not fully undo the effects of the equivalent `importaddress` call when the `p2sh` argument is true, and a copy of the script is still left behind in `mapScripts` and in the database file. It would be good to either note this as a limitation of the RPC in the documentation, or to fix it by actually removing the script and scriptid from the database.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15129#discussion_r700431854

> a copy of the script is still left behind in `mapScripts` and in the database file

Might be good follow up with @achow101 about his earlier comment here https://github.com/bitcoin/bitcoin/pull/15129#issuecomment-642769639 since I'm assuming leaving data behind in mapscripts is wasteful but doesn't cause real problems, but his comment seems to imply otherwise.

Also on Andrew's other comment about the keypool, https://github.com/bitcoin/bitcoin/pull/15129#issuecomment-643013777, I think you could address it again with documentation just saying `removeaddress` only undoes the effects of `importaddress`, it doesn't fully undo the effects of `importmulti` if the address was imported with other data like public keys.